### ISSUE TITLE
Formatting introduces a syntax error when encountering an... unconventional case/operator combo

### ISSUE THREAD

**Issue by garyb** [natefaubion/purescript-tidy#15]
``` purescript
case val of
  SE.Structure step → Path.CurStep (TDCur.cons step cur)
  SE.Type ty → Path.ProjStep (TDP.AtType ty cur)
`NEL.cons`
  ( maybe
      resPath
      (flip append resPath ∘ unfoldCursor)
      cur
  ) <> NEL.singleton Path.Root
```

Gets reformatted as:
``` purescript
case val of
  SE.Structure step → Path.CurStep (TDCur.cons step cur)
  SE.Type ty → Path.ProjStep (TDP.AtType ty cur)
  `NEL.cons`
    ( maybe
        resPath
        (flip append resPath ∘ unfoldCursor)
        cur
    ) <> NEL.singleton Path.Root
```

**Comment by garyb** [comment]
Looks like it's specifically to do with infix operators as I found a similar case where $ is used at the location of `NEL.cons` above and it doesn't rewrite that incorrectly.

**Comment by natefaubion** [comment]
Good find! 😆 I think I've got a case that handles this for applications, but must not have done it with operators.
```purescript
wat =
  case _ of
    Foo -> Bar
  42
```

**Comment by natefaubion** [comment]
It definitely formats it with an indent for operators. Maybe it's an issue with the parser then? I know there's a case to close layout blocks when encountering operators at the same column, but it just doesn't have a case for infix names.
```purescript
wat =
  case _ of
    Foo -> Bar
    $ ok
```

**Comment by garyb** [comment]
Ah! I'll go back and find the case that I though was working okay and paste it here.

**Comment by garyb** [comment]
```purescript
_userError =
  Lens.prism'
    case _ of
      AccessDeniedError ade → V.inj _accessDenied ade
      MalformedResourceError mfe → V.inj _malformedResource mfe
      WebsocketCloseCode code → review (_socketDataError' ∘ _userSocketCode) code
    $ V.on _accessDenied (Just ∘ AccessDeniedError)
    $ V.on _malformedResource (Just ∘ MalformedResourceError)
    $ (map WebsocketCloseCode <$> preview (_socketDataError' ∘ _userSocketCode))
```
🤯

**Comment by natefaubion** [comment]
Yeah, that looks correct to me, as the lhs of the `$` is an application of Lens.prism' and the case.

**Comment by garyb** [comment]
Parse issue with the other case then?

**Comment by natefaubion** [comment]
Operators use `offsideEndP`:
https://github.com/natefaubion/purescript-language-cst-parser/blob/2b3c78a8af318df29570e354116db8de7fce9ad8/src/PureScript/CST/Layout.purs#L315

Infix names use `offsideP` via `insertDefault`.
https://github.com/natefaubion/purescript-language-cst-parser/blob/2b3c78a8af318df29570e354116db8de7fce9ad8/src/PureScript/CST/Layout.purs#L266

I think this is technically a bug in both parsers, but honestly the original formatting is probably better.

**Comment by garyb** [comment]
Well, the revised formatting is a syntax error, so yes I would say the original formatting is better :smile: or is that not what you meant?

Should the CST even admit invalid syntax?

**Comment by natefaubion** [comment]
> Well, the revised formatting is a syntax error, so yes I would say the original formatting is better 😄 or is that not what you meant?

I mean that the way it formats it _should_ parse correctly. It's a bug that the parsers use a different offside rule for operators vs infix names. However, even though it _should_ parse correctly, the original formatting probably _looks_ better regardless.

**Comment by garyb** [comment]
Ah, I see! Yeah, that's very true.

**Comment by natefaubion** [comment]
> Should the CST even admit invalid syntax?

The CST supports recovery (except for lexer errors), and will emit error nodes with the recovered tokens. The formatter will still format an invalid file, but it will just print out error spans as-is.

**Comment by natefaubion** [comment]
Both the official parser and the tooling parser have a fix for the layout issue.
